ssh/tailssh/tailssh.go

```diff
@@ -91,35 +121,54 @@ func (srv *server) handleSSH(s ssh.Session) {
 		return
 	}
 
-	ptyReq, winCh, isPty := s.Pty()
-	if !isPty {
-		fmt.Fprintf(s, "TODO scp etc")
+	pol, ok := srv.sshPolicy()
+	if !ok {
+		logf("tsshd: rejecting connection; no SSH policy")
 		s.Exit(1)
 		return
 	}
+
+	ptyReq, winCh, isPty := s.Pty()
 	srcIPP := netaddr.IPPortFrom(tanetaddr, uint16(ta.Port))
 	node, uprof, ok := lb.WhoIs(srcIPP)
 	if !ok {
 		fmt.Fprintf(s, "Hello, %v. I don't know who you are.\n", srcIPP)
-		s.Exit(0)
+		s.Exit(1)
 		return
 	}
-	allow := envknob.String("TS_SSH_ALLOW_LOGIN")
-	if allow == "" || uprof.LoginName != allow {
-		logf("ssh: access denied for %q (only allowing %q)", uprof.LoginName, allow)
-		jnode, _ := json.Marshal(node)
-		jprof, _ := json.Marshal(uprof)
-		fmt.Fprintf(s, "Access denied.\n\nYou are node: %s\n\nYour profile: %s\n\nYou wanted %+v\n", jnode, jprof, ptyReq)
+
+	srcIP := srcIPP.IP()
+	sctx := &sshContext{
+		now:     time.Now(),
+		sshUser: s.User(),
+		srcIP:   srcIP,
+		node:    node,
+		uprof:   &uprof,
+	}
+	action, localUser, ok := evalSSHPolicy(pol, sctx)
+	if ok && action.Message != "" {
+		io.WriteString(s, action.Message)
+	}
+	if !ok || action.Reject {
+		logf("ssh: access denied for %q from %v", uprof.LoginName, srcIP)
 		s.Exit(1)
 		return
 	}
+	if !action.Accept || action.HoldAndDelegate != "" {
+		fmt.Fprintf(s, "TODO: other SSHAction outcomes")
+		s.Exit(1)
 
+	}
+	if !isPty {
+		fmt.Fprintf(s, "TODO scp etc\n")
+		s.Exit(1)
+		return
+	}
 	var cmd *exec.Cmd
-	sshUser := s.User()
-	if os.Getuid() != 0 || sshUser == "root" {
+	if os.Getuid() != 0 || localUser == "root" {
 		cmd = exec.Command("/bin/bash")
```

Comment:
Can we detect the login shell instead of always trying bash?

Reply:
Sure, but that's unrelated to this PR.